src/librustc/hir/lowering.rs

```diff
@@ -1871,6 +1934,237 @@ impl<'a> LoweringContext<'a> {
+                        // We don't currently have a reliable way to desugar `async fn` with
+                        // multiple potentially unrelated input lifetimes into
+                        // `-> impl Trait + 'lt`, so we report an error in this case.
+                        if current_lt_name != name {
+                            struct_span_err!(
+                                self.context.sess,
+                                current_lt_span.between(lifetime.span),
+                                E0703,
+                                "multiple different lifetimes used in arguments of `async fn`",
+                            )
+                                .span_label(current_lt_span, "first lifetime here")
+                                .span_label(lifetime.span, "different lifetime here")
+                                .help("`async fn` can only accept borrowed values \
+                                      with identical lifetimes")
+                                .emit()
+                        } else if current_lt_name.is_elided() && name.is_elided() {
+                            struct_span_err!(
+                                self.context.sess,
+                                current_lt_span.between(lifetime.span),
+                                E0704,
+                                "multiple elided lifetimes used in arguments of `async fn`",
+                            )
+                                .span_label(current_lt_span, "first lifetime here")
+                                .span_label(lifetime.span, "different lifetime here")
+                                .help("consider giving these arguments named lifetimes")
+                                .emit()
+                        }
+                    } else {
+                        self.output_lifetime = Some((name, lifetime.span));
+                    }
+                }
+            }
+        }
+
+        let bound_lifetime = {
+            let mut lifetime_collector = AsyncFnLifetimeCollector {
+                context: self,
+                currently_bound_lifetimes: Vec::new(),
+                collect_elided_lifetimes: true,
+                output_lifetime: None,
+            };
+
+            for arg in inputs {
+                hir::intravisit::walk_ty(&mut lifetime_collector, arg);
+            }
+            lifetime_collector.output_lifetime
+        };
+
+        let output_ty_name_owned;
+        let (output_ty_name, span) = match output {
+            FunctionRetTy::Ty(ty) => {
+                output_ty_name_owned = pprust::ty_to_string(ty);
+                (&*output_ty_name_owned, ty.span)
+            },
+            FunctionRetTy::Default(span) => ("()", *span),
+        };
+
+        // FIXME(cramertj) add lifetimes (see FIXME below) to the name
+        let exist_ty_name = Symbol::intern(&format!("impl Future<Output = {}>", output_ty_name));
```

Comment:
Giving it a name doesn't seem immediately useful to me, and could end up causing strange side-effects. Where can this name end up?

Reply:
It's needed to provide a name for the `hir::Item` for the existential type definition. It shouldn't ever be surfaced to the user AFAIK, but it's still useful to have a correct-ish name for debugging.

Reply:
FWIW the normal existential impl trait path does a similar thing-- it pretty prints the type to get the name.

Reply:
Doesn't it pretty-print *the bounds*?

Reply:
Replaced both with `keywords::Invalid.name()`.